File: app/src/main/java/com/fazemeright/chatbotmetcs622/ui/registration/RegistrationActivity.java
package com.fazemeright.chatbotmetcs622.ui.registration;

import android.content.Intent;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import com.fazemeright.chatbotmetcs622.R;
import com.fazemeright.chatbotmetcs622.ui.base.BaseActivity;
import com.fazemeright.chatbotmetcs622.ui.landing.LandingActivity;
import com.fazemeright.chatbotmetcs622.ui.login.LoginActivity;
import com.fazemeright.chatbotmetcs622.utils.AppUtils;

public class RegistrationActivity extends BaseActivity<RegistrationActivityViewModel>
    implements View.OnClickListener {

  private EditText userEmailEditText,
      userPasswordEditText,
      userConPasswordEditText,
      etFirstName,
      etLastName;
  private TextView tvHaveAccount;
  private Button btnRegister;

  @Override
  protected Class<RegistrationActivityViewModel> getViewModelClass() {
    return RegistrationActivityViewModel.class;
  }

  @Override
  public void initViews() {
    setUpSupportActionBar();

    userEmailEditText = findViewById(R.id.userLoginEmailEditText);
    etFirstName = findViewById(R.id.etFirstName);
    etLastName = findViewById(R.id.etLastName);
    userPasswordEditText = findViewById(R.id.userPasswordEditText);
    userConPasswordEditText = findViewById(R.id.userConPasswordEditText);
    tvHaveAccount = findViewById(R.id.tvHaveAccount);
    btnRegister = findViewById(R.id.btnRegister);

    viewModel.userRegistered.observe(this, userRegistered -> {
      if (userRegistered) {
        btnRegister.setText(getString(R.string.registration_success_msg));
        openLandingActivity();
      }
    });
  }

  private void setUpSupportActionBar() {
    if (getSupportActionBar() != null) {
      getSupportActionBar().setTitle(getString(R.string.registration));
    }
  }

  @Override
  public void setListeners() {
    btnRegister.setOnClickListener(this);
    tvHaveAccount.setOnClickListener(this);
  }

  @Override
  public int getLayoutResId() {
    return R.layout.activity_registration;
  }

  @Override
  public void onClick(View v) {
    int id = v.getId();
    if (id == R.id.tvHaveAccount) {
      openLoginActivity();
    } else if (id == R.id.btnRegister) {
      disableButton(btnRegister);
      registerUser();
      enableButton(btnRegister);
    }
  }

  private void registerUser() {
    String email = userEmailEditText.getText().toString();
    String firstName = etFirstName.getText().toString();
    String lastName = etLastName.getText().toString();
    String password = userPasswordEditText.getText().toString();
    String conPassword = userConPasswordEditText.getText().toString();
    performRegistration(email, firstName, lastName, password, conPassword);
  }

  /**
   * Open Login Activity
   */
  private void openLoginActivity() {
    startActivity(new Intent(RegistrationActivity.this, LoginActivity.class));
  }

  /**
   * Call to perform validation on the input parameters and then perform registration
   *
   * @param email       user email address
   * @param firstName   first name of user
   * @param lastName    last name of user
   * @param password    user selected password
   * @param conPassword user selected confirmation password
   */
  private void performRegistration(
      final String email,
      String firstName,
      String lastName,
      final String password,
      String conPassword) {
    if (!AppUtils.isValidEmail(email)) {
      userEmailEditText.setError(mContext.getString(R.string.incorrect_email_err_msg));
      userEmailEditText.requestFocus();
      return;
    }

    if (!AppUtils.isValidName(firstName)) {
      etFirstName.setError(mContext.getString(R.string.incorrect_first_name));
      etFirstName.requestFocus();
      return;
    }

    if (!AppUtils.isValidName(lastName)) {
      etLastName.setError(mContext.getString(R.string.incorrect_last_name));
      etLastName.requestFocus();
      return;
    }

    if (!AppUtils.isValidPassword(password)) {
      userPasswordEditText.setError(mContext.getString(R.string.incorrect_pass_err_msg));
      userPasswordEditText.requestFocus();
      return;
    }

    if (!AppUtils.arePasswordsValid(password, conPassword)) {
      userPasswordEditText.setError(mContext.getString(R.string.passwords_dont_match_err_msg));
      userPasswordEditText.requestFocus();
      userPasswordEditText.setText("");
      userConPasswordEditText.setText("");
      return;
    }

    viewModel.registerNewUser(
        email,
        password,
        firstName,
        lastName);
  }

  /**
   * Open LandingActivity and finish this one
   */
  private void openLandingActivity() {
    startActivity(new Intent(RegistrationActivity.this, LandingActivity.class));
    finish();
  }
}

File: app/src/main/java/com/fazemeright/chatbotmetcs622/ui/registration/RegistrationActivity.kt
package com.fazemeright.chatbotmetcs622.ui.registration

import android.content.Intent
import android.view.View
import com.fazemeright.chatbotmetcs622.R
import com.fazemeright.chatbotmetcs622.databinding.ActivityRegistrationBinding
import com.fazemeright.chatbotmetcs622.ui.base.BaseActivity
import com.fazemeright.chatbotmetcs622.ui.landing.LandingActivity
import com.fazemeright.chatbotmetcs622.ui.login.LoginActivity
import com.fazemeright.chatbotmetcs622.utils.AppUtils.arePasswordsValid
import com.fazemeright.chatbotmetcs622.utils.AppUtils.isValidEmail
import com.fazemeright.chatbotmetcs622.utils.AppUtils.isValidName
import com.fazemeright.chatbotmetcs622.utils.AppUtils.isValidPassword
import com.fazemeright.library.api.result.Result

class RegistrationActivity : BaseActivity<RegistrationActivityViewModel, ActivityRegistrationBinding>(), View.OnClickListener {
    override val viewModelClass: RegistrationActivityViewModel
        get() = RegistrationActivityViewModel(application)

    override fun initViews() {
        setUpSupportActionBar()

        viewModel.userRegistered.observe(this, { userRegistered: Result<Boolean> ->
            if (userRegistered is Result.Success) {
                binding.btnRegister.text = getString(R.string.registration_success_msg)
                openLandingActivity()
            }
        })
    }

    /**
     * Set up support action bar.
     */
    private fun setUpSupportActionBar() {
        supportActionBar?.apply {
            title = getString(R.string.registration)
        }
    }

    override fun setListeners() {
        binding.btnRegister.setOnClickListener(this)
        binding.tvHaveAccount.setOnClickListener(this)
    }

    override fun onClick(v: View) {
        val id = v.id
        if (id == R.id.tvHaveAccount) {
            openLoginActivity()
        } else if (id == R.id.btnRegister) {
            disableButton(binding.btnRegister)
            registerUser()
            enableButton(binding.btnRegister)
        }
    }

    /**
     * Register the user.
     */
    private fun registerUser() {
        val email = binding.userLoginEmailEditText.text.toString()
        val firstName = binding.etFirstName.text.toString()
        val lastName = binding.etLastName.text.toString()
        val password = binding.userPasswordEditText.text.toString()
        val conPassword = binding.userConPasswordEditText.text.toString()
        performRegistration(email, firstName, lastName, password, conPassword)
    }

    /**
     * Open Login Activity.
     */
    private fun openLoginActivity() {
        startActivity(Intent(this@RegistrationActivity, LoginActivity::class.java))
    }

    /**
     * Call to perform validation on the input parameters and then perform registration.
     *
     * @param email       user email address
     * @param firstName   first name of user
     * @param lastName    last name of user
     * @param password    user selected password
     * @param conPassword user selected confirmation password
     */
    private fun performRegistration(
            email: String,
            firstName: String,
            lastName: String,
            password: String,
            conPassword: String) {
        // TODO: Move to ViewModel
        if (!isValidEmail(email)) {
            binding.userLoginEmailEditText.error = context.getString(R.string.incorrect_email_err_msg)
            binding.userLoginEmailEditText.requestFocus()
            return
        }
        if (!isValidName(firstName)) {
            binding.etFirstName.error = context.getString(R.string.incorrect_first_name)
            binding.etFirstName.requestFocus()
            return
        }
        if (!isValidName(lastName)) {
            binding.etLastName.error = context.getString(R.string.incorrect_last_name)
            binding.etLastName.requestFocus()
            return
        }
        if (!isValidPassword(password)) {
            binding.userPasswordEditText.error = context.getString(R.string.incorrect_pass_err_msg)
            binding.userPasswordEditText.requestFocus()
            return
        }
        if (!arePasswordsValid(password, conPassword)) {
            binding.userPasswordEditText.error = context.getString(R.string.passwords_dont_match_err_msg)
            binding.userPasswordEditText.requestFocus()
            binding.userPasswordEditText.setText("")
            binding.userConPasswordEditText.setText("")
            return
        }
        viewModel.registerNewUser(
                email,
                password,
                firstName,
                lastName)
    }

    /**
     * Open LandingActivity and finish this one.
     */
    private fun openLandingActivity() {
        startActivity(Intent(this@RegistrationActivity, LandingActivity::class.java))
        finish()
    }

    override fun inflateLayoutFromBinding(): ActivityRegistrationBinding {
        return ActivityRegistrationBinding.inflate(layoutInflater)
    }
}